pkg/porter/dependencies.go

```diff
@@ -0,0 +1,204 @@
+package porter
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/deislabs/cnab-go/bundle"
+	"github.com/deislabs/porter/pkg/cnab/extensions"
+	cnabprovider "github.com/deislabs/porter/pkg/cnab/provider"
+	"github.com/deislabs/porter/pkg/config"
+	"github.com/deislabs/porter/pkg/context"
+	"github.com/pkg/errors"
+)
+
+type dependencyExecutioner struct {
+	*context.Context
+	Resolver BundleResolver
+	CNAB     CNABProvider
+	Manifest *config.Manifest
+
+	// These are populated by Prepare, call it or perish in inevitable errors
+	parentOpts BundleLifecycleOpts
+	action     cnabAction
+	deps       []*queuedDependency
+}
+
+func newDependencyExecutioner(p *Porter) *dependencyExecutioner {
+	resolver := BundleResolver{
+		Cache:    p.Cache,
+		Registry: p.Registry,
+	}
+	return &dependencyExecutioner{
+		Context:  p.Context,
+		Resolver: resolver,
+		CNAB:     p.CNAB,
+		Manifest: p.Manifest,
+	}
+}
+
+type cnabAction func(cnabprovider.ActionArguments) error
+
+type queuedDependency struct {
+	extensions.DependencyLock
+	CNABFile   string
+	Parameters map[string]string
+}
+
+func (e *dependencyExecutioner) Prepare(parentOpts BundleLifecycleOpts, action cnabAction) error {
+	e.parentOpts = parentOpts
+	e.action = action
+
+	err := e.identifyDependencies()
+	if err != nil {
+		return err
+	}
+
+	for _, dep := range e.deps {
+		err := e.prepareDependency(dep)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (e *dependencyExecutioner) Execute() error {
```

Comment:
Should we add a check here to see if `Prepare` needs calling?  (To avoid persishing!)  Perhaps return an error if `e.parentOpts` or `e.action` ain't valid?